contracts/core/EntryPoint.sol

```diff
@@ -397,29 +396,52 @@ contract EntryPoint is IEntryPoint, StakeManager {
     }
 
     /**
-     * revert if either account deadline or paymaster deadline is expired
+     * revert if either account sigTimeRange or paymaster sigTimeRange is expired
      */
-    function _validateDeadline(uint256 opIndex, UserOpInfo memory opInfo, uint256 deadline, uint256 paymasterDeadline) internal view {
-        //we want to treat "zero" as "maxint", so we subtract one, ignoring underflow
-    unchecked {
-        // solhint-disable-next-line not-rely-on-time
-        if (deadline != 0 && deadline < block.timestamp) {
-            if (deadline == SIG_VALIDATION_FAILED) {
-                revert FailedOp(opIndex, address(0), "AA24 signature error");
-            } else {
-                revert FailedOp(opIndex, address(0), "AA22 expired");
-            }
+    function _validateSigTimeRange(uint256 opIndex, UserOpInfo memory opInfo, uint256 sigTimeRange, uint256 paymasterTimeRange) internal view {
+        (bool sigFailed, bool outOfTimeRange) = _getSigTimeRange(sigTimeRange);
+        if (sigFailed) {
+            revert FailedOp(opIndex, address(0), "AA24 signature error");
         }
-        // solhint-disable-next-line not-rely-on-time
-        if (paymasterDeadline != 0 && paymasterDeadline < block.timestamp) {
-            address paymaster = opInfo.mUserOp.paymaster;
-            if (paymasterDeadline == SIG_VALIDATION_FAILED) {
-                revert FailedOp(opIndex, paymaster, "AA34 signature error");
-            } else {
-                revert FailedOp(opIndex, paymaster, "AA32 paymaster expired");
-            }
+        if (outOfTimeRange) {
+            revert FailedOp(opIndex, address(0), "AA22 expired or not due");
+        }
+        (sigFailed, outOfTimeRange) = _getSigTimeRange(paymasterTimeRange);
+        if (sigFailed) {
+            revert FailedOp(opIndex, opInfo.mUserOp.paymaster, "AA34 signature error");
         }
+        if (outOfTimeRange) {
+            revert FailedOp(opIndex, opInfo.mUserOp.paymaster, "AA32 paymaster expired or not due");
+        }
+    }
+
+    function _getSigTimeRange(uint sigTimeRange) internal view returns (bool sigFailed, bool outOfTimeRange) {
+        if (sigTimeRange == 0) {
+            return (false, false);
+        }
+        uint validFrom;
+        uint validUntil;
+        (sigFailed, validFrom, validUntil) = _parseSigTimeRange(sigTimeRange);
+        // solhint-disable-next-line not-rely-on-time
+        outOfTimeRange = block.timestamp > validUntil || block.timestamp < validFrom;
+    }
+
+    //extract sigFailed, validFrom, validUntil.
+    // also convert zero validUntil to type(uint64).max
+    function _parseSigTimeRange(uint sigTimeRange) internal pure returns (bool sigFailed, uint64 validFrom, uint64 validUntil) {
+        sigFailed = uint8(sigTimeRange) != 0;
+        // subtract one, to explicitly treat zero as max-value
+        validUntil = uint64(int64(int(sigTimeRange >> 8) - 1));
```

Comment:
Nice casting trick to hide the underflow without "unchecked" :)
Maybe one day Solidity will start catching casts of negative ints to uints, and our tests will catch it.